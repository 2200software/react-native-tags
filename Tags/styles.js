import { StyleSheet } from "react-native";

export default StyleSheet.create({
  container: {
  },
  scrollContainer: {
    margin: 2,
  },
  textInputContainer: {
    minWidth: 60,
    height: 32,
    margin: 2
  },

  textInput: {
    margin: 0,
    padding: 0,
    paddingLeft: 12,
    paddingRight: 12,
    flex: 1,
    height: 32,
    fontSize: 13,
    color: "rgba(0, 0, 0, 0.87)",
    borderColor: 'grey',
    borderWidth: 1,
    borderRadius: 20
  },

  tag: {
    justifyContent: "center",
    backgroundColor: "#e0e0e0",
    borderRadius: 16,
    paddingLeft: 12,
    paddingRight: 12,
    height: 32,
    margin: 4
  },
  tagLabel: {
    fontSize: 13,
    color: "rgba(0, 0, 0, 0.87)"
  }
});
