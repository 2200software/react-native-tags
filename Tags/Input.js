import React from "react";
import { View, TextInput } from "react-native";

import styles from "./styles";

const Input = (props) => {

  const {
    inputStyle,
    inputContainerStyle,
    textInputProps
  } = props;

  return (
    <View style={[styles.textInputContainer, inputContainerStyle]}>
      <TextInput
        {...textInputProps}
        style={[styles.textInput, inputStyle]}

        value={props.value}
        onChangeText={props.onChangeText}
        onSubmitEditing={props.onSubmitEditing}
        
        underlineColorAndroid="transparent"
      />
    </View>
  );
};

export {Input};
export default Input;