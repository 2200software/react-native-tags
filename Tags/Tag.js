import React from "react";
import PropTypes from "prop-types";
import { View, Text, TextInput, TouchableOpacity } from "react-native";

import styles from "./styles";

const Tag = ({ label, onPress, tagContainerStyle, tagTextStyle, readonly }) => (
  <TouchableOpacity style={[styles.tag, tagContainerStyle]} onPress={onPress} activeOpacity={readonly ? 1 : 0.2 }>
    <Text style={[styles.tagLabel, tagTextStyle]}>{label}</Text>
  </TouchableOpacity>
);

Tag.propTypes = {
  label: PropTypes.string.isRequired,
  onPress: PropTypes.func
};

export default Tag;
