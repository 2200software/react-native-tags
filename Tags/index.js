import React from "react";
import PropTypes from "prop-types";
import { FlatList, View } from "react-native";

import Tag from "./Tag";
import Input from "./Input";
import styles from "./styles";

class Tags extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      tags: props.initialTags,
      text: props.initialText
    };
  };

  showLastTag = () => {
    this.setState(state =>
    ({
      tags: state.tags.slice(0, -1),
      text: state.tags.slice(-1)[0] || undefined
    }),
      () =>
        this.props.onChangeTags && this.props.onChangeTags(this.state.tags)
    );
  };

  addTag = text => {
    this.setState(state =>
    ({
      tags: [...state.tags, text.trim()],
      text: undefined
    }),
      () => this.props.onChangeTags && this.props.onChangeTags(this.state.tags)
    );
  };

  onChangeText = text => {
    if (text.length === undefined) {
      this.showLastTag();
    } else if (
      text.length > 1 &&
      this.props.createTagOnString.includes(text.slice(-1)) &&
      !text.match(new RegExp(`^[${this.props.createTagOnString.join("")}]+$`, 'g')) &&
      !(this.state.tags.indexOf(text.slice(0, -1).trim()) > -1)
    ) {
      this.addTag(text.slice(0, -1));
    } else {
      this.setState({ text });
    }
  };

  onSubmitEditing = () => {
    if (!this.props.createTagOnReturn) {
      return;
    }
    this.addTag(this.state.text);
  };

  render() {

    const {
      containerStyle,
      style,
      readonly,
      maxNumberOfTags,
      tagContainerStyle,
      tagTextStyle,
      deleteTagOnPress,
      onTagPress,
      renderTag
    } = this.props;

    return (
      <View style={[styles.container, containerStyle, style]}>
        <FlatList
          numColumns={maxNumberOfTags}
          columnWrapperStyle={{ flexWrap: 'wrap', margin: 2 }}
          horizontal={false}
          showsVerticalScrollIndicator={true}
          style={[styles.scrollContainer, style]}
          data={this.state.tags}
          renderItem={({ item, index }) => {
            const tagProps = {
              tag: item,
              index,
              deleteTagOnPress,
              onPress: event => {
                event?.persist();
                if (deleteTagOnPress && !readonly) {
                  this.setState(state =>
                  ({
                    tags: [
                      ...state.tags.slice(0, index),
                      ...state.tags.slice(index + 1)
                    ]
                  }),
                    () => {
                      this.props.onChangeTags &&
                        this.props.onChangeTags(this.state.tags);
                      onTagPress && onTagPress(index, item, event, true);
                    }
                  );
                } else {
                  onTagPress && onTagPress(index, item, event, false);
                }
              },
              tagContainerStyle,
              tagTextStyle
            };
            return renderTag(tagProps);
          }}
        />

        {!readonly
          && maxNumberOfTags > this.state.tags.length
          &&
          <Input
            value={this.state.text}
            onChangeText={this.onChangeText}
            onSubmitEditing={this.onSubmitEditing}
            {...this.props}
          />
        }

      </View>
    );
  };

}

Tags.defaultProps = {
  initialTags: [],
  initialText: undefined,
  createTagOnString: [",", " "],
  createTagOnReturn: false,
  readonly: false,
  deleteTagOnPress: true,
  maxNumberOfTags: Number.POSITIVE_INFINITY,
  renderTag: ({ tag, index, ...rest }) => (
    <Tag key={`${tag}-${index}`} label={tag} {...rest} />
  )
};

Tags.propTypes = {
  initialText: PropTypes.string,
  initialTags: PropTypes.arrayOf(PropTypes.string),
  createTagOnString: PropTypes.array,
  createTagOnReturn: PropTypes.bool,
  onChangeTags: PropTypes.func,
  readonly: PropTypes.bool,
  maxNumberOfTags: PropTypes.number,
  deleteTagOnPress: PropTypes.bool,
  renderTag: PropTypes.func,
  /* style props */
  containerStyle: PropTypes.any,
  style: PropTypes.any,
  inputContainerStyle: PropTypes.any,
  inputStyle: PropTypes.any,
  tagContainerStyle: PropTypes.any,
  tagTextStyle: PropTypes.any,
  textInputProps: PropTypes.object
};

export { Tags };
export default Tags;
