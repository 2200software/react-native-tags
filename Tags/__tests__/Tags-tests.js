import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Tags from '../../';


describe('Tags', () => {
  function createComponent(props = {}) {
    return (
      <Tags
        initialText="monkey gland sauce."
        initialTags={['tomato sauce', 'mustard', 'mayo']}
      />
    )
  }
  

  it('should render props correctly', () => {
    const tree = renderer.create(createComponent()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // TODO: Test that:
  // should create a new tag and clear the text input when ` ` or `,` are entered.
  // should remove the last tag, and populate the text input when backspace is entered.
  // should call `onChangeTags` when tags are added and removed.

});
