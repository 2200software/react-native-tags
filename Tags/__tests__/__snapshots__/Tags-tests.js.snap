exports[`Tags should render props correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "wrap"
      }
    ]
  }>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e0e0e0",
          "borderRadius": 16,
          "height": 32,
          "justifyContent": "center",
          "margin": 4,
          "paddingLeft": 12,
          "paddingRight": 12
        }
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13
          }
        ]
      }>
      tomato sauce
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e0e0e0",
          "borderRadius": 16,
          "height": 32,
          "justifyContent": "center",
          "margin": 4,
          "paddingLeft": 12,
          "paddingRight": 12
        }
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13
          }
        ]
      }>
      mustard
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#e0e0e0",
          "borderRadius": 16,
          "height": 32,
          "justifyContent": "center",
          "margin": 4,
          "paddingLeft": 12,
          "paddingRight": 12
        }
      ]
    }>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontSize": 13
          }
        ]
      }>
      mayo
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#ccc",
          "borderRadius": 16,
          "flex": 1,
          "height": 32,
          "margin": 4,
          "width": 100
        }
      ]
    }>
    <TextInput
      onChangeText={[Function bound onChangeText]}
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "flex": 1,
            "fontSize": 13,
            "height": 32,
            "margin": 0,
            "padding": 0,
            "paddingLeft": 12,
            "paddingRight": 12
          }
        ]
      }
      underlineColorAndroid="transparent"
      value="monkey gland sauce." />
  </View>
</View>
`;
